### ISSUE TITLE
JsonTextReader number parsing performance

### ISSUE THREAD

**Issue by JamesNK** [JamesNK/Newtonsoft.Json#772]
JsonTextReader parses characters twice when reading numbers.

First it reads number characters into a buffer - https://github.com/JamesNK/Newtonsoft.Json/blob/865c08d95d89565b4f6b463b57da8f5324f6ce7c/Src/Newtonsoft.Json/JsonTextReader.cs#L1653 - then it attempts to convert the buffer into the desired type (int/long/double/decimal). For integers it can do this without turning the buffer into a string, for decimal and double the characters are converted to a string and double.Parse or decimal.Parse are called.

Goal is to only read characters once and to avoid ever allocating strings.

e.g. read number chars into a long. If a comma is encountered then the long is returned. If a decimal point is encountered a new long value is read. When a comma is encountered those two longs are then used to build a double/decimal, etc.

**Comment by BrutalSimplicity** [comment]
I am interested in working on this, but am having trouble even building the solution. Are there any guidelines to help contributors get started?

**Comment by JamesNK** [comment]
There are some things that need to happen before a single pass parsing would be possible. First there needs to be helper methods created that can parse double and decimal values directly from a char array. Only once those are done then a method that combines them together with an existing method that can parse integers.

**Comment by BrutalSimplicity** [comment]
Yes, I saw the double parsing issue after posting to this. It seems that there is a pretty good starting point there. I'll look at those areas and see if there is a way for me to contribute.

Right now, it looks like it's close, but there are precision issues. Correct?

**Comment by TylerBrinkley** [comment]
This can be closed now can't it or is there anything else that needs to be done?

**Comment by JamesNK** [comment]
My end vision was to be able to parse numbers in one iteration over the characters. Right now there are still two: one to read chars until the end of the number (fetching more from the TextReader if required), and one to then convert those chars to an int/decimal/double/etc.

I'm not sure how feasible it will be to get it down to one pass, especially taking into account sync vs async.
